doc/oomd/SONiC OOM daemon.md

```diff
@@ -0,0 +1,267 @@
+    - Can set allow list for user/group:
+       - Only process run by these users/groups can be terminate by OOMD.
+       - This setting is designed to protect key system process.
+    - Can set terminate policy:
+       - Terminate user session or terminate user process.
+       - The reason to terminate user session are:
+         - When kernel receive a allocate memory request, kernel will try find memory by memory swap or release memory used by cache.
+         - SONiC not enable swap, when swap not enabled, the kernel will finish the find memory procedure very fast.
+         - When set a very high high-water mark, if OOMD configured to terminate process, OOMD may can't terminate process quickly enough to release memory.
+         - Terminate user session will block user from running any new command, will provide better memory protection in this case.
+- Provide default setting to minimize customer side change.
+  - Customers may have network management software to initialize device configuration, because this feature add new commands, those management software may need update to support new commands. The default limitation is designed to cover most scenario to minimize the customer side change.
+
+# 2 Configuration and Management Requirements
+## 2.1 SONiC CLI
+ - Manage login session or memory  limit settings
+```
+    # config OOMD enable/disable status
+    config oomd { enable/disable}
+    
+    # config oomd policy
+    config oomd policy { policyname } { add|del } <value>
+```
+ - Show oomd config
+```
+    show oomd policy
+    show oomd status
+```
+
+## 2.2 Config DB
+ - OOMD feature can be enable/disable by config DB
+ - OOMD policy are fully configurable by config DB.
+
+# 3 Design
+ - High-level design diagram:
+
+<img src="./High-level design diagram.PNG"/>
+
+## 3.1 OOMD Implementation
+ - OOMD will be a service managed by systemd.
+ - System memory monitor and protect
+     - Get system memory information from /proc/meminfo.
+     - Subscript cgroup memory utilization event from /sys/fs/cgroup/memory/memory.usage_in_bytes.
+     - Get user information from /etc/passwd.
+     - OOMD will send SIGKILL/SIGTERM to kill/terminate process.
+     - OOMD can config to terminate user session by kill/terminate process group.
+ - Docker container monitor and re-start
+     - Subscript docker container OOM event from memory.oom_control
+     - OOMD will run 'docker restart' command to restart docker container
+ - The following diagram show how OOMD work:
+    - UserInfoProvider will get user information from /etc/passwd
+    - ConfigProvider will subscript config DB change event and update OOMD config.
+    - MemMonitor will monitor cgroup event and trigger OomHandler.
+    - ContainerMonitor will monitor container OOM event and re-start container
+    - When memory utilization reach high-water mark, OomHandler will terminate user process to prevent OOM happen. 
```

Comment:
>happen [](http://example.com/codeflow?start=107&length=6)

In the new PNG graph, there is 'ConfigProvider'. we don't has such kind of concept, and I think the downstream could directly subscribe/read ConfigDB by a common library `swsscommon`.

Reply:
Fixed, the PNG updated.